libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -238,64 +243,166 @@ string IRGenerator::generateInitialAssignment(VariableDeclaration const& _varDec
+					params += (params.empty() ? "" : ", ") + m_context.addLocalVariable(*varDecl).commaSeparatedList();
+			construtorFunction("params", params);
+			string baseParamsString = listAllParams(baseConstructorParams);
+			construtorFunction("baseParams", baseParamsString);
+			construtorFunction("comma", !params.empty() && !baseParamsString.empty() ? ", " : "");
+			construtorFunction("functionName", constructorName(*contract));
+			pair<string, map<ContractDefinition const*, string>> evaluatedArgs = evaluateConstructorArguments(*contract);
+			baseConstructorParams.insert(evaluatedArgs.second.begin(), evaluatedArgs.second.end());
+			construtorFunction("initBaseArguments", evaluatedArgs.first);
+			construtorFunction("hasNextConstructor", i < _contract.annotation().linearizedBaseContracts.size() - 1);
+			if (i < _contract.annotation().linearizedBaseContracts.size() - 1)
+			{
+				ContractDefinition const* nextContract = _contract.annotation().linearizedBaseContracts[i + 1];
+				construtorFunction("nextConstructor",constructorName(*nextContract));
+				construtorFunction("nextParams", listAllParams(baseConstructorParams, nextContract));
+			}
+			construtorFunction("initStateVariables", initStateVariables(*contract));
+			FunctionDefinition const* constructor = contract->constructor();
+			construtorFunction("hasUserDefinedConstructor", constructor);
+			if (constructor)
+				construtorFunction("constructorBody", generate(constructor->body()));
 
-		Whiskers t(R"X(
-			let programSize := datasize("<object>")
-			let argSize := sub(codesize(), programSize)
+			return construtorFunction.render();
+		});
+	}
+
+	return constructorsCode;
+}
 
-			let memoryDataOffset := <allocate>(argSize)
-			codecopy(memoryDataOffset, programSize, argSize)
+string IRGenerator::constructorCode(ContractDefinition const& _contract)
+{
+	solAssert(!_contract.isLibrary(), "Tried to initialize state variables of library.");
 
-			<assignToParams> <abiDecode>(memoryDataOffset, add(memoryDataOffset, argSize))
+	Whiskers creationCode(R"(
+		<?hasCallValueCheck>
+			<callValueCheck>
+		</hasCallValueCheck>
 
-			<constructorName>(<params>)
-		)X");
-		t("object", creationObjectName(_contract));
-		t("allocate", m_utils.allocationFunction());
-		t("assignToParams", paramVars == 0 ? "" : "let " + suffixedVariableNameList("param_", 0, paramVars) + " := ");
-		t("params", suffixedVariableNameList("param_", 0, paramVars));
-		t("abiDecode", abiFunctions.tupleDecoder(constructor->functionType(false)->parameterTypes(), true));
-		t("constructorName", m_context.enqueueFunctionForCodeGeneration(*constructor));
+		<?hasParams>
+			let <params> := <copyArgumentsToMemory>()
+		</hasParams>
+		<constructor>(<params>)
+	)");
+	FunctionDefinition const* constructor = _contract.constructor();
+	creationCode("hasCallValueCheck", constructor && !constructor->isPayable());
+	creationCode("callValueCheck", callValueCheck());
```

Comment:
Maybe remove the `<?hasCallValueCheck>` and just use an empty string here if it does not have a call value check?

Reply:
Makes sense, did like that